src/Compilers/Core/Portable/DiagnosticAnalyzer/DefaultAnalyzerAssemblyLoader.Core.cs

```diff
@@ -2,45 +2,138 @@
+            lock (_guard)
             {
-                AssemblyLoadContext loadContext = AssemblyLoadContext.GetLoadContext(typeof(DefaultAnalyzerAssemblyLoader).GetTypeInfo().Assembly);
+                if (!_loadContextByDirectory.TryGetValue(fullDirectoryPath, out loadContext))
+                {
+                    loadContext = new DirectoryLoadContext(fullDirectoryPath, this);
+                    _loadContextByDirectory[fullDirectoryPath] = loadContext;
+                }
+            }
+
+            var name = AssemblyName.GetAssemblyName(fullPath);
+            return loadContext.LoadFromAssemblyName(name);
+        }
 
-                if (System.Threading.Interlocked.CompareExchange(ref _loadContext, loadContext, null) == null)
+        private bool ShouldLoadInAnalyzerContext(string? assemblySimpleName, string fullPath)
+        {
+            foreach (var compilerAssemblySimpleName in CompilerAssemblySimpleNames)
+            {
+                if (string.Equals(compilerAssemblySimpleName, assemblySimpleName, StringComparison.OrdinalIgnoreCase))
                 {
-                    _loadContext.Resolving += (context, name) =>
-                    {
-                        Debug.Assert(ReferenceEquals(context, _loadContext));
-                        return Load(name.FullName);
-                    };
+                    return false;
                 }
             }
 
-            return LoadImpl(fullPath);
+            return IsKnownDependencyLocation(fullPath);
         }
 
-        protected virtual Assembly LoadImpl(string fullPath) => _loadContext.LoadFromAssemblyPath(fullPath);
+        private sealed class DirectoryLoadContext : AssemblyLoadContext
+        {
+            private readonly string _directory;
+            private readonly DefaultAnalyzerAssemblyLoader _loader;
+
+            public DirectoryLoadContext(string directory, DefaultAnalyzerAssemblyLoader loader)
+            {
+                _directory = directory;
+                _loader = loader;
+            }
+
+            protected override Assembly? Load(AssemblyName assemblyName)
+            {
+                var simpleName = assemblyName.Name;
+                var assemblyPath = Path.Combine(_directory, simpleName + ".dll");
+                if (!_loader.ShouldLoadInAnalyzerContext(simpleName, assemblyPath))
+                {
+                    return null;
+                }
+
+                var pathToLoad = _loader.GetPathToLoad(assemblyPath);
+                try
+                {
+                    return LoadFromAssemblyPath(pathToLoad);
+                }
+                catch (FileNotFoundException)
+                {
+                    return null;
+                }
+            }
+
+            protected override IntPtr LoadUnmanagedDll(string unmanagedDllName)
+            {
+                var assemblyPath = Path.Combine(_directory, unmanagedDllName + ".dll");
```

Comment:
Shouldn't this go through `_loader.GetPathToLoad`?

Reply:
Fixed